src/mod_webdav.c

```diff
@@ -1792,20 +1809,56 @@ static handler_t mod_webdav_put(server *srv, connection *con, plugin_data *p, ha
 				case ENOENT:
 					con->http_status = 404; /* not found */
 					break;
+				case EISDIR:
+					/* RFC4918 Section 9.7.2 */
+					con->http_status = 405; /* method not allowed */
+					break;
+				case ELOOP:
+				case EACCES:
+					con->http_status = 403; /* forbidden */
+					break;
 				default:
-					con->http_status = 403; /* not found */
+					/* potential bug */
+					con->http_status = 500;
 					break;
 				}
 				return HANDLER_FINISHED;
 			}
 
-			if (-1 == lseek(fd, offset, SEEK_SET)) {
-				con->http_status = 501; /* not implemented */
+			do {
+				if (complete_length >= 0) {
+					off_t file_length = lseek(fd, 0, SEEK_END);
+					if (file_length < 0) break;
+					if (complete_length < file_length) {
+						/* RFC7233 Section 4.4 */
```

Comment:
You mention RFC 7233 Section 4.4 here, but I do not think the changes in this commit are valid.  Would you please explain the problem you thought you were trying to solved with "improve Content-Range for PUT"?  What webdav client are you using which is sending Content-Range with PUT?  RFC 7231 rejects Content-Range with PUT.
Please see https://tools.ietf.org/html/rfc7231#section-4.3.4 :
>    An origin server that allows PUT on a given target resource MUST send
>    a 400 (Bad Request) response to a PUT request that contains a
>    Content-Range header field (Section 4.2 of [RFC7233]), since the
>    payload is likely to be partial content that has been mistakenly PUT
>    as a full representation.

Reply:
PUT with Content-Range is a popular extension for webdav since it enables editing a file just like doing it locally. See https://github.com/miquels/webdavfs/blob/master/DAV-requirements.md#for-writing-files , also Finder the OS X's built-in WebDAV client uses it.

Reply:
Since the client indicates the length of the resource, it must know it and if it's wrong, the server should warn the client instead of silently commiting changes. But if you are not satisfied with this part, I don't care about this particular behavior and it can be removed.

Reply:
> But if you are not satisfied with this part

TBH, I am not satisfied with your cavalier approach to RFCs.  Please remove this patch, which is a **direct violation** of RFC7231.

https://tools.ietf.org/html/rfc7231#section-4.3.4 is very explicit about the required behavior.  Absent any errata that you can point me to which relaxes that requirement (and which I did not find), the MUST is a MUST.  That section in the RFC to which I linked follows with a pointers to how one might properly do a partial file update:
>    Partial content updates are possible by
>    targeting a separately identified resource with state that overlaps a
>    portion of the larger resource, or by using a different method that
>    has been specifically defined for partial updates (for example, the
>    PATCH method defined in [RFC5789]).

PATCH is not currently implemented in lighttpd mod_webdav and might be one proper approach to supporting this feature.  (I have a larger set of changes to mod_webdav which I plan to push in the next week or so, so if you choose to look into implementing PATCH, the base might change, so please hold off for now.)

Reply:
OK, I'll wait for those changes before I adjust my patches.